compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -392,6 +430,46 @@ static void setAlwaysVectorizeSizes(linalg::LinalgOp op,
   }
 }
 
+static void setVectorSizesForDynamicShapes(
+    linalg::LinalgOp op, SmallVectorImpl<int64_t> &parallelSizes,
+    SmallVectorImpl<int64_t> &reductionSizes) {
+  SmallVector<int64_t> origParallelSizes(parallelSizes.begin(),
+                                         parallelSizes.end());
+  SmallVector<int64_t> origReductionSizes(reductionSizes.begin(),
+                                          reductionSizes.end());
+  setAlwaysVectorizeSizes(op, parallelSizes, reductionSizes);
+
+  // If peeling is enabled and the 'op' is fully dynamic, we only vectorize the
+  // lowest order parallel dimension for now to avoid peeling higher level
+  // dimensions. If no parallel dimension is found to be vectorized, we try to
+  // vectorize the lowest order reduction dimension.
+  if (isFullyDynamicOp(op) && isPeelingBeneficial(op)) {
+    bool isParallelDimVectorized = false;
+    for (int i = origParallelSizes.size() - 1; i >= 0; --i) {
+      if (origParallelSizes[i] > 1) {
+        assert(parallelSizes[i] == 1 &&
+               "This tile size should have been set to one");
+        parallelSizes[i] = origParallelSizes[i];
+        isParallelDimVectorized = true;
+        break;
+      }
+    }
+
+    if (!isParallelDimVectorized) {
+      for (int i = origReductionSizes.size() - 1; i >= 0; --i) {
+        if (origReductionSizes[i] > 1) {
+          assert(reductionSizes[i] == 1 &&
+                 "This tile size should have been set to one");
+          reductionSizes[i] = origReductionSizes[i];
+          break;
+        }
+      }
+    }
+  }
+
+  return;
```

Comment:
Let's make the level of nest lower. We can exit directly when it does not meet the requirements. E.g.,

```
if (!isFullyDynamicOp(op) || !isPeelingBeneficial(op)) return;
// Check isParallelDimVectorized...
if (isParallelDimVectorized) return;
// Check reduction cases..
```